lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -2138,88 +2163,142 @@ void FIRRTLLowering::emitRandomizePrologIfNeeded() {
 }
 
 void FIRRTLLowering::initializeRegister(Value reg) {
+  typedef std::pair<Attribute, std::pair<unsigned, unsigned>> SymbolAndRange;
+
+  // The point in the design where we should add randomization register
+  // definitions.  This is at the top of the "`ifndef SYNTHESIS" block.
+  mlir::OpBuilder::InsertPoint regInsertionPoint;
+
+  auto regDef = cast<sv::RegOp>(reg.getDefiningOp());
+  if (!regDef->hasAttrOfType<StringAttr>("inner_sym"))
+    regDef->setAttr("inner_sym", builder.getStringAttr(moduleNamespace.newName(
+                                     Twine("__") + regDef.name() + "__")));
+  auto regDefSym =
+      hw::InnerRefAttr::get(theModule.getNameAttr(), regDef.inner_symAttr());
+
   // Construct and return a new reference to `RANDOM.  It is always a 32-bit
   // unsigned expression.  Calls to $random have side effects, so we use
   // VerbatimExprSEOp.
   constexpr unsigned randomWidth = 32;
-  auto getRandom32Val = [&]() -> Value {
-    return builder.create<sv::VerbatimExprSEOp>(
-        builder.getIntegerType(randomWidth), "`RANDOM");
-  };
-
-  // Return an expression containing random bits of the specified width.
-  // An explicit std::function is required here due to recursion.
-  std::function<Value(IntegerType)> getRandomValue =
-      [&](IntegerType type) -> Value {
-    assert(type.getWidth() != 0 && "zero bit width's not supported");
-    auto &randomValueAndRemain = blockRandomValueAndRemain[builder.getBlock()];
-    if (randomValueAndRemain.second >= type.getWidth()) {
-      auto value = builder.createOrFold<comb::ExtractOp>(
-          type, randomValueAndRemain.first,
-          randomWidth - randomValueAndRemain.second);
-
-      randomValueAndRemain.second -= type.getWidth();
-      return value;
+  auto getRandom32Val = [&](Twine suffix = "") -> Value {
+    sv::RegOp randReg;
+    {
+      OpBuilder::InsertionGuard topBuilder(builder);
+      builder.restoreInsertionPoint(regInsertionPoint);
+      randReg = builder.create<sv::RegOp>(
+          reg.getLoc(), builder.getIntegerType(randomWidth),
+          /*name=*/builder.getStringAttr("_RANDOM"),
+          /*inner_sym=*/
+          builder.getStringAttr(moduleNamespace.newName(Twine("_RANDOM"))));
     }
```

Comment:
This might be an extremely dumb idea, but could we reuse the `randReg` of a previous calls to `getRandom32Val`? Like adding at most one such reg to any given module, and then just reusing that as a temporary to write a new ``` `RANDOM ``` into? I don't think we ever use the value in that `randReg` once we've used up all bits.

Reply:
This is an interesting idea that I didn't consider... 🤔 This would result in fewer (one) temporary variables. There would be more procedural assigns for types larger than 32-bits, though I'm not sure that matters.

```verilog
_RANDOM_0 = `RANDOM;
_RANDOM_1 = `RANDOM;
foo <= {_RANDOM_0, _RANDOM_1};
```

Would become:
```verilog
_RANDOM = `RANDOM;
foo[31:0] <= _RANDOM;
_RANDOM = `RANDOM;
foo[63:32] <= _RANDOM;
```

This unrolled loop could be rolled up for style reasons (if tools will accept variable part select assigns?).